sage.py

```diff
@@ -1343,105 +95,78 @@ def load_IANA_mapping():
                 }
-        else:
-            # Do nothing 
-            pass
     return ports
-    
+
+
 ## 3
 def readfile(fname):
-    unparsed_data = None
     with open(fname, 'r') as f:
         unparsed_data = json.load(f)
         
     unparsed_data = unparsed_data[::-1]
-    #print('# events: ', len(unparsed_data))
-    #print(unparsed_data[0])
     return unparsed_data
     
 
-#cats = dict()
-#ips = dict()
-#hosts = dict()
-#h_trig = []
-
 def parse(unparsed_data, alert_labels=[], slim=False):
-    
+
     FILTER = False
     badIP = '169.254.169.254'
     __cats = set()
     __ips = set()
     __hosts = set()
-    __sev = set()
-    data = []
-    connections = dict()
+    parsed_data = []
 
     prev = -1
-    for id, d in enumerate(unparsed_data):
-        #print(d)
-        raw = ''
-        
-        try:
-             raw = json.loads(d['result']['_raw'])
-        except:
-            try:
-                raw = json.loads(d['_raw'])
-            except:
-                raw = d
-                
+    for d in unparsed_data:
+        if 'result' in d and '_raw' in d['result']:
+            raw = json.loads(d['result']['_raw'])
+        elif '_raw' in d:
+            raw = json.loads(d['_raw'])
+        else:
+            raw = d
+
         if raw['event_type'] != 'alert':
             continue
-        #app_proto = raw['app_proto']
-        host = ''
-        
-        try:
+
+        if 'host' in raw:
             host = raw['host']
-        except:
-            try:
-                host = d['host'][3:]
-            except:
-                host = 'dummy'
-                
+        elif 'host' in d:
+            host = d['host'][3:]
+        else:
+            host = 'dummy'
 
-        #print(host)
         ts = raw['timestamp']
-        dt = datetime.datetime.strptime(ts, '%Y-%m-%dT%H:%M:%S.%f%z')# 2018-11-03T23:16:09.148520+0000
-        DIFF = 0.0 if prev== -1 else round((dt - prev).total_seconds(),2)
+        dt = datetime.datetime.strptime(ts, '%Y-%m-%dT%H:%M:%S.%f%z')  # 2018-11-03T23:16:09.148520+0000
+        diff_dt = 0.0 if prev == -1 else round((dt - prev).total_seconds(), 2)
         prev = dt
-        
-                
+
         sig = raw['alert']['signature']
         cat = raw['alert']['category']
-        
-        severity = raw['alert']['severity']
 
+        # Filter out the alert that occurs way too often
         if cat == 'Attempted Information Leak' and FILTER:
             continue
+
         srcip = raw['src_ip']
         srcport = None if 'src_port' not in raw.keys() else raw['src_port']
         dstip = raw['dest_ip']
         dstport = None if 'dest_port' not in raw.keys() else raw['dest_port']
 
-        # Filtering out mistaken alerts / uninteresting alerts
+        # Filter out mistaken alerts / uninteresting alerts
         if srcip == badIP or dstip == badIP or cat == 'Not Suspicious Traffic':
             continue
-            
+
         if not slim:
```

Comment:
Is this `slim` part still useful? Some methods below (e.g. `remove_duplicates`) rely on 9th field of an alert, which is `mcat`. Using `slim` will result in an exception

Reply:
slim is set to false I think. Basically, the alert is parsed and then the get_attack_stage_mapping is called to get the attack stage that matches the alert signature. This is done for all cases

Reply:
The badIP, "Not suspicious traffic", "attempted info leak" are all edge cases meant for the experimental datasets. A general user does not necessarily have to run this part of the code. For now I think we can keep it in, and leave a comment that this can be removed if it is not needed.

Reply:
okay, so I see that alert_labels is always empty and because slim is set to false, the code around "Not suspicious traffic" and "attempted info leak" is basically unreachable. So I think the whole block (if slim:) can be removed.

Reply:
Alright, I will remove the `if slim:` block. As for the `badIP` etc parts in the main part of the method, I don't think that the average user will be searching for the hints like that in the methods, so maybe it would be better to comment them out or at least mention it somewhere so that the user knows that those parts are being used?